cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/RuntimeClientService.java

```diff
@@ -96,8 +117,29 @@ public class RuntimeClientService extends AbstractRetryableScheduledService {
     this.runtimeClient = runtimeClient;
     this.fetchLimit = cConf.getInt(Constants.RuntimeMonitor.BATCH_SIZE);
     this.programFinishTime = new AtomicLong(-1L);
+
+    // Initialize encoders and decoders for metrics aggregation.
+    Schema schema = null;
+    DatumWriter<MetricValues> recordWriter = null;
+    DatumReader<MetricValues> reader = null;
+    try {
+      TypeToken<MetricValues> metricValueType = TypeToken.of(MetricValues.class);
+      schema = new ReflectionSchemaGenerator().generate(metricValueType.getType());
+      recordWriter = new ASMDatumWriterFactory(new ASMFieldAccessorFactory()).create(metricValueType, schema);
+      reader = new ReflectionDatumReaderFactory().create(metricValueType, schema);
+    } catch (UnsupportedTypeException e) {
+       if (cConf.getBoolean(Constants.RuntimeMonitor.METRICS_AGGREGATION_ENABLED)) {
```

Comment:
Can't we move METRICS_AGGREGATION_ENABLED to the top. Having this check inside the catch clause makes it a bit strange.

Reply:
Changed.